nemo/collections/tts/models/magpietts.py

```diff
@@ -567,79 +589,156 @@ def forward(self, dec_input_embedded, dec_input_mask, cond, cond_mask, attn_prio
-            all_preds.append(codebook_preds)
-
-        all_preds = torch.stack(all_preds, dim=1)  # (B, C, T')
+        all_preds = [[] for _ in range(self.frame_stacking_factor)]
+        for fs_index in range(self.frame_stacking_factor):
+            for idx in range(self.num_audio_codebooks):
+                si = (idx + self.num_audio_codebooks * fs_index) * self.num_all_tokens_per_codebook
+                ei = si + self.num_all_tokens_per_codebook
+                codebook_logits = all_code_logits[:, :, si:ei]
+                codebook_probs = torch.softmax(codebook_logits, dim=-1)  # (B, T', num_tokens_per_codebook)
+                # argmax to get the tokens
+                codebook_preds = torch.argmax(codebook_probs, dim=-1)  # (B, T')
+                all_preds[fs_index].append(codebook_preds)
+        all_preds = [torch.stack(p, dim=1) for p in all_preds] # list of `frame_stacking_factor`` elements of shape (B,C,T) each
+        all_preds = torch.stack(all_preds, dim=-1) # B, C, T, frame_stacking_factor
+        # undo the frame stacking
+        all_preds = all_preds.reshape(all_preds.size(0), all_preds.size(1), -1) # B, C, T*frame_stacking_factor
+        pred_max_len = all_preds.size(2)
+        real_max_len = audio_codes_lens.max()
+        assert (pred_max_len - real_max_len) < self.frame_stacking_factor
+        # trim padding introduced for frame stacking
+        all_preds = all_preds[:,:, :real_max_len]
         audio_mask = get_mask_from_lengths(audio_codes_lens)
         all_preds = all_preds * audio_mask.unsqueeze(1)
 
         return all_preds
 
-    def local_transformer_sample_maskgit(self, dec_output, temperature=0.7, topk=80, unfinished_items={}, finished_items={}, use_cfg=False, cfg_scale=1.0, n_steps=3):
+    def debug_visualize_codes(self,codes, mask_id=2020, frame_stacking_rate=2):
         """
-        Sample codes for one timestep from the local transformer using MaskGit.
+        Visualize codes for debug purposes
+        codes: (B, C)
         """
+        def code_to_str(code):
+            if code==mask_id:
+                return "M    "
+            else:
+                return f"{code:04d} "
+        DS = frame_stacking_rate
+        B, C = codes.shape
+        if B > 1:
+            print("Warning: visualizing only first batch element")
+        codes = codes.clone().detach().cpu().numpy()[0]
+        codes = [code_to_str(c) for c in codes]
+        for i, c in enumerate(codes):
+            if (i) % (C/frame_stacking_rate) == 0:
+                print("  |timestep| ", end="")
+            print(c, end="")
+        print()
+
+    def get_forbidden_tokens(self) -> list[int]:
+        """
+        Returns a list of tokens that should not be sampled.
+        """
+        # Note that audio EOS is not included because we *do* need to sample it at the end of the utterance.
+        return [self.mask_token_id, self.audio_bos_id, self.context_audio_bos_id, self.context_audio_eos_id]
```

Comment:
What about the reserved tokens? Alternatively, if EOS is the only one we care about. Maybe we should have an allowlist instead

Reply:
Alternatively, we could define this in the SpecialAudioToken enum itself